### PR TITLE
Loosen nullability restrictions added in #3205 (#3226)

### LINKED ISSUE #3226: new_null_array Panics creating StructArray with non-nullable fields
**Describe the bug**

Following #3205 new_null_array panics if creating a StructArray with non-nullable children

**To Reproduce**

**Expected behavior**

**Additional context**

### PR DESCRIPTION
**Pull request by tustvold**
# Which issue does this PR close?

Closes #3226

# Rationale for this change
 

#3205 added restrictions to prevent nulls in non-nullable children of a StructArray. Whilst on the face of it this sounds sensible, the issue is that nulls in StructArray and FixedSizeListArray both take up space in their children. This leads to the question of "what value should the child have for a null in its parent". In the case of DictionaryArray there isn't a valid placeholder value that can be used, the value at a null index is undefined.

This PR works around this by loosening the restriction to allow a null in a non-nullable child array provided it corresponds to a null in the parent. This ensures that the children are valid, and can be safely interpreted outside the context of their parent, whilst also ensuring that the semantics of nullability are preserved.

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

File: arrow-array/src/array/mod.rs [hunk 1]
```diff
@@ -916,7 +916,7 @@ mod tests {
     #[test]
     fn test_null_struct() {
         let struct_type =
-            DataType::Struct(vec![Field::new("data", DataType::Int64, true)]);
+            DataType::Struct(vec![Field::new("data", DataType::Int64, false)]);
```
Comment (tustvold) [inline arrow-array/src/array/mod.rs:919]: This reverts a change in #3205 - you should be able to create a null struct containing a non-nullable child

Comment (alamb) [reply arrow-array/src/array/mod.rs:919]:
```suggestion
            // It is possible to create a null struct containing a non-nullable child 🤯 
            // see https://github.com/apache/arrow-rs/pull/3244 for details
            DataType::Struct(vec![Field::new("data", DataType::Int64, false)]);
```

It might be a good idea to leave some rationale for this very non obvious corner case

File: arrow-array/src/array/string_array.rs [hunk 2]
```diff
@@ -609,7 +609,7 @@ mod tests {
 
         let offsets = [0, 5, 10].map(|n| O::from_usize(n).unwrap());
         let data_type = GenericListArray::<O>::DATA_TYPE_CONSTRUCTOR(Box::new(
-            Field::new("item", DataType::UInt8, false),
+            Field::new("item", DataType::UInt8, true),
```
Comment (alamb) [inline arrow-array/src/array/string_array.rs:612]:
```suggestion
            // It is possible to create a null struct containing a non-nullable child 🤯 
            // see https://github.com/apache/arrow-rs/pull/3244 for details
            Field::new("item", DataType::UInt8, true),
```

File: arrow-array/src/array/struct_array.rs [hunk 3]
```diff
@@ -568,7 +562,9 @@ mod tests {
     }
 
     #[test]
-    #[should_panic(expected = "non-nullable field cannot have null values")]
+    #[should_panic(
+        expected = "non-nullable child of type Int32 contains nulls not present in parent Struct"
```
Comment (alamb) [inline arrow-array/src/array/struct_array.rs:566]: 👍

File: arrow-data/src/data.rs [hunk 4]
```diff
@@ -1012,9 +1020,100 @@ impl ArrayData {
...
+                                }
+                                for j in 0..element_len {
+                                    bit_util::set_bit(
+                                        buffer.as_mut(),
+                                        i * element_len + j,
+                                    )
+                                }
+                            }
+                            let mask = buffer.into();
+                            self.validate_non_nullable(Some(&mask), 0, child)?;
```
Comment (alamb) [inline arrow-data/src/data.rs:1055]: I think some comments about this mask (what it is and why it is built) might be helpful

File: arrow-data/src/data.rs [hunk 5]
```diff
@@ -1012,9 +1020,100 @@ impl ArrayData {
...
+        mask: Option<&Buffer>,
+        offset: usize,
+        data: &ArrayData,
+    ) -> Result<(), ArrowError> {
+        let mask = match mask {
+            Some(mask) => mask.as_ref(),
+            None => return match data.null_count {
+                0 => Ok(()),
+                _ => Err(ArrowError::InvalidArgumentError(format!(
+                    "non-nullable child of type {} contains nulls not present in parent {}",
```
Comment (alamb) [inline arrow-data/src/data.rs:1086]: I suggest making these two messages different so it is easier to find what error is being hit by `grep` if someone hits the error

**Review by alamb — APPROVED** [review]
LGTM -- cc @Jefffrey

**Review by Jefffrey — APPROVED** [review]
Thanks for this, looks like there was some nuance I missed in regards to nested types & nullability 👍

**Comment by tustvold** [comment]
> Thanks for this, looks like there was some nuance I missed in regards to nested types & nullability

Not at all, the ecosystem as a whole does not seem to handle this in a consistent manner. It certainly isn't documented :laughing:
